### PR TITLE
put converters in their place

### PR DESCRIPTION
**Pull request by mdehoon**
and get rid of _BinaryFormats. Let's see if this works.

- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file, have run ``flake8`` locally, and
understand that AppVeyor and TravisCI will be used to confirm the Biopython unit
tests and style checks pass with these changes.

- [X] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

### CODE REVIEW

File: Tests/test_SeqIO_convert.py [hunk 1]
```diff
@@ -199,8 +198,43 @@ def failure_check(self, filename, in_format, out_format, alphabet):
...
+    ("fastq", "fastq-illumina"),
+    ("fastq-sanger", "fastq-illumina"),
+    ("fastq-solexa", "fastq-illumina"),
+    ("fastq-illumina", "fastq-illumina"),
+    ("fastq", "qual"),
+    ("fastq-sanger", "qual"),
+    ("fastq-solexa", "qual"),
+    ("fastq-illumina", "qual"),
+)
+
```
Comment (peterjc) [inline Tests/test_SeqIO_convert.py]: I prefer automatically testing all the combinations over having a hard coded list which must be updated by hand when something new is added.

File: Bio/SeqIO/InsdcIO.py [hunk 2]
```diff
@@ -1519,6 +1519,21 @@ class ImgtWriter(EmblWriter):
     FEATURE_HEADER = "FH   Key                 Location/Qualifiers\nFH\n"
 
 
+def convert(in_stream, in_format, out_stream, out_format):
```
Comment (peterjc) [inline Bio/SeqIO/InsdcIO.py]: If you want this to be private (as per the docstring comment and prior usage), it needs a leading underscore.

**Comment by peterjc** [comment]
Do you strongly prefer removing `` Bio/SeqIO/_convert.py`` and putting any optimised converters elsewhere (for now just InsdcIO and QualityIO)?

File: Bio/SeqIO/InsdcIO.py (hunk 2 above)
Comment (mdehoon) [reply Bio/SeqIO/InsdcIO.py]: It cannot be private; this function is called from `SeqIO/__init__.py`

**Comment by mdehoon** [comment]
> Do you strongly prefer removing ` Bio/SeqIO/_convert.py` and putting any optimised converters elsewhere (for now just InsdcIO and QualityIO)?

Yes, because otherwise you'll end up sharing private variables across modules. Also, putting converters in a `_convert.py` module means dividing related code across modules. For example, we define `FastqGeneralIterator` in `Bio.SeqIO.QualityIO` but then use it in `Bio.SeqIO._convert`. Some day we will change something in `Bio.SeqIO.QualityIO`and forget that `Bio.SeqIO._convert` also relies on it. And finally, a `_convert.py` module will end up with lots of unrelated code in the same module (e.g. `GenBankScanner` and `FastqGeneralIterator`).

File: Bio/SeqIO/InsdcIO.py (hunk 2 above)
Comment (peterjc) [reply Bio/SeqIO/InsdcIO.py]: The public function ``Bio.SeqIO.convert(...)`` can call a private function ``Bio.SeqIO.InsdcIO._convert(...)`` - that isn't a problem.

**Comment by peterjc** [comment]
Happily the conversion code has reasonable unit test coverage, so that shouldn't happen either way.

**Comment by mdehoon** [comment]
@peterjc  OK to merge? The `test_SeqIO*` tests can still be cleaned up further, but that can happen in a separate PR.

**Comment by peterjc** [comment]
You haven't fixed the mismatch with ``Bio.SeqIO.InsdcIO.convert(...)`` with no leading underscore, and its docstring saying private (because it used to be in a private module).

I would probably go for ``Bio.SeqIO.InsdcIO._convert(...)`` here, leaving it private.

**Comment by mdehoon** [comment]
@peterjc  I don't think we should allow access to private functions from other modules in Biopython. This will quickly become a mess, in particular also because different modules in Biopython are maintained by different people who may not have a good overview of how other modules are designed.

**Comment by peterjc** [comment]
In that case make as a public function ``Bio.SeqIO.InsdcIO.convert(...)`` needs a better docstring. And some more user friendly behaviour like a ValueError when the output is not "fasta" etc. Returning None only made sense in the context of the private API it was written for.

**Comment by mdehoon** [comment]
@peterjc  This function is still not meant to be used by end-users.

**Comment by peterjc** [comment]
Yes  - which is why it was private in the first place, and why I suggested keeping it private by adding a leading underscore now that you have moved it out of the private module.

**Comment by mdehoon** [comment]
@peterjc  There are three categories of functions:
1) Functions meant to be called by end-users;
2) Functions meant to be called by other modules in Biopython, but not by end-users;
3) Functions meant to be called only from within the module in which it is defined.

If we add a leading underscore, but still call the function from a different modules, then we are not distinguishing between 2 and 3, which IMHO is essential for encapsulation and clean code design.

**Comment by peterjc** [comment]
Maybe case 2.5, all under ``Bio/SeqIO/`` but not in the same file.

**Comment by mdehoon** [comment]
@peterjc 
For case 2.5 I would still not add an underscore. If we did, then we'd be giving the message that it is OK to call private functions from other files. We could create some rule that this is only allowed within the same top-level module, but generally developers (especially new developers, or users trying to fix a bug) may not be aware of this rule, and compliance will be poor. In addition, in some cases it's not obvious what counts as the same module, e.g. if two modules are highly dependent on each other (e.g. Bio.SeqRecord, Bio.SeqIO, and Bio.SeqFeature, or Bio.Align and Bio.AlignIO).

**Comment by peterjc** [comment]
I didn't see an ideal solution back when I came up with the single private module ``_convert.py`` in the first place as a place to put all the optimisation code.

If you _really_ want to drop that module, and want public functions, I think two public functions in ``InsdcIO.py`` made sense:

```python
def genbank_to_fasta(in_stream, out_stream):
    """Optimised function called by Bio.SeqIO.convert internally.

    While you can call this function directly, you are encouraged to use
    Bio.SeqIO.convert(in_stream, "genbank", out_stream, "fasta")
    instead.

    Returns the number of records.
    """
    records = GenBankScanner().parse_records(in_stream, do_features=False)
    return SeqIO.write(records, out_stream, "fasta")

def embl_to_fasta(in_stream, out_stream):
    """Optimised function called by Bio.SeqIO.convert internally.

    While you can call this function directly, you are encouraged to use
    Bio.SeqIO.convert(in_stream, "embl", output_filename, "fasta")
    instead.

    Returns the number of records.
    """
    records = EmblScanner().parse_records(in_stream, do_features=False)
    return SeqIO.write(records, out_stream, "fasta")
```

Likewise in ``Bio.SeqIO.QualityIO``, you seem happy with the private function ``_fastq_illumina_convert_fastq_sanger(...)`` etc but that could be public as ``fastq_illumina_to_sanger(...)`` instead?

**Comment by mdehoon** [comment]
@peterjc  But how would `Bio.SeqIO.convert` then know which conversions are supported by `Bio.SeqIO.InsdcIO`? Then `Bio.SeqIO.__init__.py` would have to contain a list of supported conversions, and the exact name of each conversion function, and you're back to square one.

**Comment by peterjc** [comment]
Exactly - this is a core design choice of ``Bio.SeqIO``, a registry of parsers and writers (done as dictionaries keyed by format), with the handful of optimised converters keyed on a pair of formats.

I see no reason to change that - wasn't you goal here just to reorganise a private sub-module?

**Comment by mdehoon** [comment]
@peterjc 
> I see no reason to change that - wasn't you goal here just to reorganise a private sub-module?

My goal is to avoid private variables getting shared across modules.

> Exactly - this is a core design choice of Bio.SeqIO, a registry of parsers and writers (done as dictionaries keyed by format), with the handful of optimised converters keyed on a pair of formats.

Then we should do exactly that. Right now we have multiple registries:
- `_FormatToIterator` in `Bio/SeqIO/__init__.py`
- `_FormatToString` in `Bio/SeqIO/__init__.py`
- `_FormatToWriter` in `Bio/SeqIO/__init__.py`
- `_BinaryFormats` in `Bio/SeqIO/__init__.py`
- `_converter` in `Bio/SeqIO/_convert.py`
- `BINARY_FORMATS` in `test_SeqIO_QualityIO.py`

You don't need this many; a single registry (`_modules` maybe) in `Bio/SeqIO/__init__.py` is sufficient, without needing to access private variables across any modules, anywhere.

**Comment by peterjc** [comment]
Just move the ``Bio/SeqIO/_convert.py`` dictionary into ``Bio/SeqIO/__init__.py`` then. That's the only place (other than perhaps tests) which needs access to it.

**Comment by peterjc** [comment]
(I don't see any to replace the dictionaries with something more object based, but if we wanted to make the format registry public that might make sense - I just kept it simple at the start)

**Comment by mdehoon** [comment]
@peterjc  To make things simpler, I have moved changes to the `test_SeqIO_FastaIO.py` tests to a separate PR (#2810 ), and removed them from this PR.

**Comment by mdehoon** [comment]
@peterjc 

> Just move the `Bio/SeqIO/_convert.py` dictionary into `Bio/SeqIO/__init__.py` then. That's the only place (other than perhaps tests) which needs access to it.

Then the dictionary in `Bio/SeqIO/__init__.py` will point to private functions in `Bio/SeqIO/_convert.py` ...

In any case, I have started to split up this PR into components that hopefully are more manageable.

**Comment by mdehoon** [comment]
Most of this PR has now been merged into master by a number of smaller PRs. Closing this PR.

**Pull request closed without merging**